### PR TITLE
LibWeb: Load external scripts in SVG

### PR DESCRIPTION
**Pull request by shlyakpavel**

### CODE REVIEW

File: Libraries/LibWeb/SVG/SVGScriptElement.cpp [hunk 1]
```diff
@@ -40,33 +42,52 @@ void SVGScriptElement::process_the_script_element()
...
+        script_url = document().parse_url(href_value);
+        if (!script_url.is_valid()) {
+            // Handle invalid IRI reference
+            dbgln("Invalid script URL: {}", href_value);
+            return;
+        }
+
+        // Fetch the external script content
+        auto request = LoadRequest::create_for_url_on_page(script_url, &document().page());
+        auto resource = ResourceLoader::the().load_resource(Resource::Type::Generic, request);
```
Comment (shannonbooth) [inline Libraries/LibWeb/SVG/SVGScriptElement.cpp:64]: This should be updated to use fetch, see: https://github.com/LadybirdBrowser/ladybird/issues/2634

Comment (shlyakpavel) [reply Libraries/LibWeb/SVG/SVGScriptElement.cpp:64]: Sounds like a right thing to do, but looks overcomplicated. I haven't found a simple few-lines way to do that (yet)

**Comment by shlyakpavel** [comment]
https://wpt.fyi/results/svg/scripted still fail as CDATA scripts
```
  <script><![CDATA[
    test(function() {
      var tspan = document.getElementById('tsp');
      assert_equals(tspan.dx.baseVal.length, 6);
      assert_equals(tspan.dy.baseVal.length, 9);
    });
  ]]></script>
  ```
  Are not supported. However, these tests now load both resources/testharness.js and resources/testharnessreport.js

File: Libraries/LibWeb/SVG/SVGScriptElement.cpp (hunk 1 above)
Comment (shlyakpavel) [reply Libraries/LibWeb/SVG/SVGScriptElement.cpp]: @shannonbooth it uses fetch now

**Review by shannonbooth — APPROVED** [review]
Looks reasonable enough to me, there are definitely some further improvements here and subtle things we would be getting wrong that could potentially be solved by some type of code sharing with HTMLScriptElement, but this is an improvement nonetheless

**Comment by shlyakpavel** [comment]
Rebased on master

**Comment by shlyakpavel** [comment]
After rebase, that doesn't build anymore on Linux, and I need tome for investigation

**Comment by shlyakpavel** [comment]
I added `IGNORE_USE_IN_ESCAPING_LAMBDA`. The reason is that we do `HTML::main_thread_event_loop().spin_until` in this function, so we know for sure lambda ends before the variable gets out of scope.

File: Libraries/LibWeb/SVG/SVGScriptElement.cpp [hunk 2]
```diff
@@ -40,15 +43,75 @@ void SVGScriptElement::process_the_script_element()
...
+    auto script_url = m_document->url();
 
-    // FIXME: 2. If the 'script' element references external script content, then the external script content
-    //           using the current value of the 'xlink:href' attribute is fetched. Further processing of the
-    //           'script' element is dependent on the external script content, and will block here until the
-    //           resource has been fetched or is determined to be an invalid IRI reference.
+    // 2. If the 'script' element references external script content, then the external script content
+    //    using the current value of the 'xlink:href' attribute is fetched. Further processing of the
+    //    'script' element is dependent on the external script content, and will block here until the
+    //    resource has been fetched or is determined to be an invalid IRI reference.
```
Comment (ADKaster) [inline Libraries/LibWeb/SVG/SVGScriptElement.cpp:52]: Goodness the spec ref at the top of this function is ancient.

We should try to follow this, I presume?

https://www.w3.org/TR/SVG2/linking.html#processingURL-parsing
https://www.w3.org/TR/SVG2/linking.html#processingURL-fetch

File: Libraries/LibWeb/SVG/SVGScriptElement.cpp [hunk 3]
```diff
@@ -40,15 +43,75 @@ void SVGScriptElement::process_the_script_element()
...
+        if (!script_url.is_valid()) {
+            dbgln("Invalid script URL: {}", href_value);
+            return;
+        }
+
+        auto& vm = realm().vm();
+        auto request = Fetch::Infrastructure::Request::create(vm);
+        request->set_url(script_url);
+        request->set_destination(Fetch::Infrastructure::Request::Destination::Script);
+        request->set_mode(Fetch::Infrastructure::Request::Mode::NoCORS);
```
Comment (ADKaster) [inline Libraries/LibWeb/SVG/SVGScriptElement.cpp:66]: > When fetching external resources from the Internet, user agents must use a [potentially CORS-enabled request](https://html.spec.whatwg.org/multipage/urls-and-fetching.html#create-a-potential-cors-request) as defined in HTML [[HTML](https://www.w3.org/TR/SVG2/refs.html#ref-html)] with the corsAttributeState as follows:
>    For an ‘[href](https://svgwg.org/specs/animations/#HrefAttribute)’ reference on an ‘[image](https://www.w3.org/TR/SVG2/embedded.html#ImageElement)’ element or ‘[script](https://www.w3.org/TR/SVG2/interact.html#ScriptElement)’ element, the CORS state specified by the ‘[crossorigin](https://www.w3.org/TR/SVG2/embedded.html#ImageElementCrossoriginAttribute)’ attribute.
    
It's fine for now to use NoCORS here, at least until Luke's PR is merged. A FIXME saying to read the `crossorigin` attribute would be appropriate.

File: Libraries/LibWeb/SVG/SVGScriptElement.cpp [hunk 4]
```diff
@@ -63,10 +126,7 @@ void SVGScriptElement::process_the_script_element()
     if (!m_document->ready_to_run_scripts())
         HTML::main_thread_event_loop().spin_until(GC::create_function(heap(), [&] { return m_document->ready_to_run_scripts(); }));
```
Comment (ADKaster) [inline Libraries/LibWeb/SVG/SVGScriptElement.cpp:128]: does this happen? 

> Note that a [load](https://www.w3.org/TR/SVGMobile12/interact.html#EventsLoad) event is dispatched on a ['script'](https://www.w3.org/TR/SVGMobile12/script.html#ScriptElement) element once it has been processed, unless it referenced external script content with an [invalid IRI reference](https://www.w3.org/TR/SVGMobile12/intro.html#TermInvalidIRIReference) and ['externalResourcesRequired'](https://www.w3.org/TR/SVGMobile12/struct.html#ExternalResourcesRequiredAttribute) was set to 'true'.

(from the SVGMobile12 spec)

If not a FIXME might be in order.

**Review by ADKaster — COMMENTED** [review]
two FIXME requests, otherwise this looks reasonable

File: Libraries/LibWeb/SVG/SVGScriptElement.cpp (hunk 4 above)
Comment (shlyakpavel) [reply Libraries/LibWeb/SVG/SVGScriptElement.cpp:128]: This does not happen. Adding a fixme

File: Libraries/LibWeb/SVG/SVGScriptElement.cpp (hunk 3 above)
Comment (shlyakpavel) [reply Libraries/LibWeb/SVG/SVGScriptElement.cpp:67]: Added a FIXME

**Review by ADKaster — APPROVED** [review]

**Review by ADKaster — APPROVED** [review]

**Pull request merged**